rust-runtime/inlineable/src/client_request_compression.rs

```diff
@@ -0,0 +1,256 @@
+    ) -> Cow<'_, RuntimeComponentsBuilder> {
+        Cow::Borrowed(&self.runtime_components)
+    }
+}
+
+#[derive(Debug)]
+struct RequestCompressionInterceptorState {
+    options: Option<CompressionOptions>,
+}
+
+impl Storable for RequestCompressionInterceptorState {
+    type Storer = StoreReplace<Self>;
+}
+
+/// Interceptor for Smithy [`@requestCompression`][spec].
+///
+/// [spec]: https://smithy.io/2.0/spec/behavior-traits.html#requestcompression-trait
+pub(crate) struct RequestCompressionInterceptor {}
+
+impl fmt::Debug for RequestCompressionInterceptor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RequestCompressionInterceptor").finish()
+    }
+}
+
+impl RequestCompressionInterceptor {
+    pub(crate) fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Intercept for RequestCompressionInterceptor {
+    fn name(&self) -> &'static str {
+        "RequestCompressionInterceptor"
+    }
+
+    fn read_before_serialization(
+        &self,
+        _context: &BeforeSerializationInterceptorContextRef<'_>,
+        _runtime_components: &RuntimeComponents,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let disable_request_compression = cfg
+            .load::<DisableRequestCompression>()
+            .cloned()
+            .unwrap_or_default();
+        let request_min_compression_size_bytes = cfg
+            .load::<RequestMinCompressionSizeBytes>()
+            .cloned()
+            .unwrap_or_default();
+        let options = CompressionOptions::default()
+            .with_min_compression_size_bytes(request_min_compression_size_bytes.0)?
+            .with_enabled(!disable_request_compression.0);
+
+        let mut layer = Layer::new("RequestCompressionInterceptor");
+        layer.store_put(RequestCompressionInterceptorState {
+            options: Some(options),
+        });
+        cfg.push_layer(layer);
+
+        Ok(())
+    }
+
+    fn modify_before_signing(
+        &self,
+        context: &mut BeforeTransmitInterceptorContextMut<'_>,
+        _runtime_components: &RuntimeComponents,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let state = cfg
+            .load::<RequestCompressionInterceptorState>()
+            .expect("set in `read_before_serialization`");
+
+        let options = state.options.clone().unwrap();
+        let request = context.request_mut();
+
+        // Don't wrap a body if compression is disabled.
+        if !options.is_enabled() {
+            tracing::trace!("request compression is disabled and will not be applied");
+            return Ok(());
+        }
+
+        // Don't wrap a body if it's below the minimum size
+        //
+        // Because compressing small amounts of data can actually increase its size,
+        // we check to see if the data is big enough to make compression worthwhile.
+        if let Some(known_size) = request.body().bytes().map(|b| b.len()) {
```

Comment:
**question**: If I"m reading `ByteStream` and `SdkBody` type implementations correctly `SdkBody.bytes()` may not work correctly for file inputs and would restrict compression to only in-memory representations. Is this intentional as it seems like a possibly undesired limitation?

Reply:
```rust
// Don't wrap a body if it's below the minimum size
//
// Because compressing small amounts of data can actually increase its size,
// we check to see if the data is big enough to make compression worthwhile.
if let Some(known_size) = request.body().bytes().map(|b| b.len()) {
    if known_size < options.min_compression_size_bytes() as usize {
        tracing::trace!("request body is below minimum size and will not be compressed");
        return Ok(());
    }
    tracing::trace!("compressing non-streaming request body...")
} else {
    tracing::trace!("compressing streaming request body...");
}
```

The reasoning is:
1. When a body doesn't have a known length and compression is enabled, then we compress that body.
2. When a body does have a known length and compression is enabled, then we compress the body only if it's above the `min_compression_size_bytes`.

In the code, we're checking for an in-memory body, which is an indirect way of checking that it's streaming vs. non-streaming. In this case, we may end up compressing files that would be below the `min_compression_size_bytes`.

Reply:
I changed this to use `SizeHint` so that sized streaming bodies will be treated the same as in-memory bodies. Streaming bodies without a length will always be compressed _(so long as request compression is enabled and appropriate for the operation.)_